test/peer-test.js

```diff
@@ -0,0 +1,74 @@
+/* eslint-env mocha */
+/* eslint prefer-arrow-callback: "off" */
+
+'use strict';
+
+const assert = require('bsert');
+const Logger = require('blgr');
+const common = require('../lib/net/common');
+const Peer = require('../lib/net/peer');
+const Network = require('../lib/protocol/network');
+const network = Network.regtest;
+
+async function sleep(time) {
+  return new Promise(resolve => setTimeout(resolve, time));
+}
+
+describe('Peer (Unit)', function() {
+  const peer = new Peer({
+    'network': network,
+    logger: new Logger({
+      level: 'debug',
+      file: false,
+      console: false
+    })
+  });
+
+  beforeEach(() => {
+    peer.banScore = 0;
+    peer.dosScore = 0;
+  });
+
+  it('should increase ban score', async () => {
+    peer.increaseBan(10);
+    assert.strictEqual(peer.banScore, 10);
+
+    peer.increaseBan(10);
+    assert.strictEqual(peer.banScore, 20);
+  });
+
+  it('should increase dos score', async () => {
+    peer.increaseDos(10, Date.now());
+    assert.strictEqual(peer.dosScore, 10);
+
+    peer.increaseDos(10, Date.now());
+    assert.strictEqual(peer.dosScore, 20);
+  });
+
+  it('should increase dos score with decay', async () => {
+    peer.increaseDos(30, Date.now());
+    assert.strictEqual(peer.dosScore, 30);
+
+    await sleep(1000);
+    // (30 * 0.5) + 10
+
+    peer.increaseDos(10, Date.now());
+    assert.strictEqual(peer.dosScore, 25);
+  });
+
+  it('should reset dos score after decay', async () => {
+    peer.increaseDos(10, Date.now());
+    assert.strictEqual(peer.dosScore, 10);
+
+    // override reset time to 0s
+    const dosResetTime = common.DOS_RESET_TIME;
+    common.DOS_RESET_TIME = 0;
```

Comment:
I want to suggest allowing a configurable dos reset time on the pool and peer objects, but I noticed the ratio of the transient dos score's lifespan to its half-life is 30/1. This ratio is held in the btcd implementation, as well. How sensitive is this ratio to manipulation? Would we need to keep this ratio constant and update the exponential function to account for a dynamic half-life?  Related, how would updating these parameters affect the precomputed decay factors? Specifically, is the size of the precomputed table a magic number, or can it be calculated dynamically?

Reply:
I'm trying to grok whether it is better to keep this value constant and only give the user one parameter to play around with (the dos score limit).

Reply:
So, we have the following equations:

    dN/dt = −λN

    N = N0e−λt

    T1/2 = 0.693/λ

Here N = max dos score, λ = decay constant and T1/2 = half-life

In our case, we want to provide rate of decay (dN/dt) as a config and tweak constant `λ` accordingly. Latest commit adds a `PeerOptions` option `p2prps`, which is a generic option for tweaking max rps peers are allowed, using which we compute `λ` and use it calculate the half-life.

I've tweaked default params such that we maintain a steady rate of 100 requests per second, but obviously callers can tweak this.

(Note: temporarily removed precomputed constants since it depends on half-life now).

Reply:
So, our scenario is a bit more complicated. In the above equations, the score `N` never increases. However for us, `N` can increase because of new requests linearly as well as decrease logarithmically because of the decay.

In practice it appears that the rps is about 10% of the configured rps. It might be possible to compute the 'practical' rps by using some computations. Not sure if it's worth the time investment. If the rps always scales 0.1x we could just use a rule of thumb and 10x the config parameter. :man_shrugging:

Reply:
OK, so the calculation is not too difficult:

At time `t` seconds, assuming an `rps` requests per second, we have incoming:

`t * rps` requests each adding 1 dos score. At the same time, assuming half-life 1s, the cumulative decay of previous scores (each rps decaying in 1s) is:

    rps / 2^1 + rps / 2^2 + rps / 2^3 ... rps / 2^t = rps / 2 ^ t

so total:

    t * rps - rps * 2 ^ t  = rps * (t - 1/2^t)

So we conclude that at any give time `t` we will have `rps * (t - 1/2^t)` dos score which is the actual or practical rps. This is the scaling factor we got above (0.1x), so we can generalize it.

Reply:
Correction: since the incoming scores are already accounted for in the decaying scores, the total score at time `t` is simply:

    rps * (1 - 1/2^t)

Reply:
So, all this means is if you are aiming to serve `rps` requests per second, when `t` is small you need to overshoot by a multiple of about `1-2x` to actually achieve the required rps. This is because recent rps will clog up more dos score when starting up. As `t` approaches infinity, you will get the required rate as all the old dos scores have decayed. I hope the explanation makes sense.

Reply:
> So, all this means is if you are aiming to serve rps requests per second, when t is small you need to overshoot by a multiple of about 1-2x to actually achieve the required rps.

@tuxcanfly are you saying that, in practice, if I want to serve `x` requests per second, but my overall decay reset period is quite small, I will need to configured `p2prps` as `2x` because small `t` result in a bloated dos score due to the nature of exponential drop offs?

Reply:
Yep, that's right